python/ray/train/v2/api/data_parallel_trainer.py

```diff
@@ -195,13 +198,29 @@ def _initialize_and_run_controller(self, **controller_init_kwargs) -> Result:
             )(TrainController)
 
             controller = controller_actor_cls.remote(**controller_init_kwargs)
+
+            self._register_sigint_handler(controller)
+
             ray.get(controller.run.remote())
             return ray.get(controller.get_result.remote())
         else:
             controller = TrainController(**controller_init_kwargs)
-            controller.run()
+            asyncio.run(controller.run())
             return controller.get_result()
 
+    def _register_sigint_handler(self, controller: TrainController):
+        """Register SIGINT handler so user Ctrl C gracefully aborts run."""
+
+        def sigint_handler(signum, frame):
+            try:
+                ray.get(controller.abort.remote())
+            except ray.exceptions.ActorDiedError:
+                # We catch the error and exit 0 to indicate graceful termination.
+                # However, for some reason the process still exits with 1.
+                sys.exit(0)
+
```

Comment:
Why do we want this? Should it just pass over the error and exit with code 130 (interrupted)?

Reply:
I have an explanation here: https://github.com/ray-project/ray/pull/53600#discussion_r2146284079. Open to suggestions though.